### ISSUE TITLE
Deal with loss of acceptance correctly

### ISSUE THREAD

**Issue by cyberphysic4l** [iotaledger/iota-core#315]

**Comment by jonastheis** [comment]
Let's take a look at what the scenarios that we are worried about actually are:

1. no validator is online because they all crashed -> **once a validator restarts it needs to start a special procedure to revive the chain**
2. no validator is online (form validator perspective) because I lost connection -> **just continue in my own partition and potentially switch chain once partitions are merged (no need for special procedure)**
3. no validator is online because a user lost connection to all of them/is offline -> **nothing to do**

Therefore, the only reason why the progress halted can be that all validators crashed. That means the last persisted state on disk for any validator is its last committed slot + potentially accepted blocks in the slots that were not yet committed. 

## Procedure to revive the network 

### 1. we commit to empty slots in between
precondition: we need to commit until slot 10 to be able to produce a slot in 14.

**flavor a:**
- complications with the commitment of the slots that we need to commit to. specifically in the example 6 and 7. We need to take all accepted blocks and just commit these slots with these blocks.
- we then need to reference all of these accepted blocks as otherwise they won't have a future cone (and can not be synced correctly)
- the new block(s) we produce in slot 14 commits to slot 10 but references parents in slot 6/7. -> the new future cone
- this has the benefit that we don't need to change anything in the way we commit / track accepted blocks
- drawback: might need to produce multiple blocks to refernce all "accepted tips" which needs special rules in CC and performance factor


**flavor b:**
- instead of committing to all the previously accepted blocks in slot 6 and 7 we only commit to those that we can reference with 1 block.
- this has the advantage that we don't need to introduce a special case for this behavior for performance factor and CC
- has the drawback that we need to unaccept blocks and rollback individual state. this might be really nasty.

![Image](https://github.com/iotaledger/iota-core/assets/4181434/c11a0ab5-502b-488a-a744-7d7821a4cafa)



### 2. backdate blocks
in this solution we simply backdate blocks and issue them in a certain pattern.
- also here we need to make sure that we reference all accepted blocks in 6 and 7 directly or indirectly.
- we could do this by starting from slot 9 and issue at the last nanosecond of that slot.
- realistically, it would be best for the validator to issue at the necessary interval of eg 1 second starting from wherever acceptance stopped to maximize rewards.
- problem: we need to make sure that the validator does not equivocate within a slot or subslot and thus forfeit its rewards -> we could simply persist the last issued validation block and after restart begin issuing from there
- it might be best to produce the full chain in secret first before sending it out as it otherwise allows users to also issue backdated blocks and with some luck these might be approved and ultimately end up accepted


![Image](https://github.com/iotaledger/iota-core/assets/4181434/b0ca92b5-ec36-44ea-bc09-e5f9b764bc4d)


## Conclusion
Solution 2, backdating blocks seems simpler to implement and has less moving parts.

However, there are still some questions:
- How does the scheduler handle this influx of validation blocks once the entire chain gets solid? We need to make sure that the blocks are forwarded as that is the only way the network can resume
- if a validator went offline/crashed before others and it "missed" some progress and commitments, currently there's no way for the validator to realize that its state is outdated. Therefore, we might need a network primitive that allows to ask your neighbors to ask for their latest commitments so that the validator can sync to that state from any node and continue from there.

**Comment by oliviasaa** [comment]
Intuitively, the tricky part for me is to make sure nobody is punished by unintentionally equivocating or something like that.

**Comment by jonastheis** [comment]
> Intuitively, the tricky part for me is to make sure nobody is punished by unintentionally equivocating or something like that.

If we persist the latest validation block issued and start backdating blocks / reviving the network only **after that latest issued block** I can't really see a way of equivocation within a subslot. 
Similarly, with that block we store which commitment/chain we voted upon and we can make sure to not issue blocks on two separate chains within the same slot. Basically, just check that the block you're going to issue in the same slot are committing to the same chain. If not, just start issuing in the next slot.

**Comment by cyberphysic4l** [comment]
> In [this document](https://docs.google.com/document/d/1trytLYpPyfvU-GaA4DCS4uhOckkl1-3Cxe1RrkPzRD8/edit#heading=h.n6g8ez53yybp), this will be playing with MAX_TOKENS, or even adding some funny conditions taking the issuance time of the blocks into consideration

I don't think MAX_TOKENS parameter can be used in this way to allow such a burst of validation blocks through. This period of time could be several hours, and this parameter is intended to absorb small burst or perhaps a few seconds due to network delays, but having it very large also allows very large bursts of spam so this is not good. 

I have a proposal which should solve this and should more generally solve the problem of "catching up" when we fall out of sync. So with the scenario from Jonas' second figure above, the validator will come back online and essentially construct this chain of red blocks approving all the neccessary accepted block from slot 6/7, creating commitments for all slots up until slot 10, and lets assume this is all done in a matter of seconds. For the validator, its own red blocks can follow the normal dataflow, except tip selection because we want the validator to select its own blocks before scheduling them. We should actually have a special rule for validation blocks that your own ones are passed to your tip manager right away rather than waiting to be scheduled. 

The acceptance of the whole chain of red blocks up until slot 14, and the corresponding commitments of the slots will happen in the dataflow without requiring anything to actually be scheduled by the validator. So I propose that we skip committed blocks in the scheduler (does not consume part of the throughput), by hooking to eviction. It makes perfect sense to skip committed blocks anyway, and they don't need to be forwarded either. 

Neighbouring nodes of the validator will first see the blocks from slot 12 or 13 because these ones are not committed by the validator so they get scheduled and gossipped as normal. They will see the new commitments and try to sync that new chain. The process will be then very similar to the validator - they will accept and commit the red blocks themselves as they go, and skip the blocks in the scheduler whenever they commit them, all the way up slot 12 which is not committed. Then the excess throughput of the validation scheduler (it schedules at a rate COMMITTEE_SIZE*VALIDATION_BLOCKS_PER_SLOT/SLOT_DURATION for each validator queue to ensure they have more throughput than needed, but penalises upon commitment if they have actually exceeded VALIDATION_BLOCKS_PER_SLOT within a slot).

This also allows us to spec the required buffer size of each validator queue quite precisely. We need VALIDATION_BLOCKS_PER_SLOT*MAX_COMMITTABLE_AGE to ensure we could have a backlog of MAX_COMMITTABLE_AGE worth of validation blocks.

**Comment by cyberphysic4l** [comment]
What we need to implement:

1. Block issuer should store the latest issued validation block and use it for tip selection somehow...
2. Procedure to backdate. Start with a set of accepted blocks that are on disc. Need to make sure that everything we end up committing has a future cone, so we need to select any accepted block without children. These accepted blocks without children may be from committed slot as well as the uncommitted ones, need to check back some number of slots in the past (probably minCA or maxCA or something.)
3. Modify the scheduler to skip committed blocks, and don't forward skipped blocks. See suggestions above about buffer sizes for the scheduler.
4. Tests..

**Comment by jonastheis** [comment]
> the idea of backdating introduces a strange logic in the protocol
It is strange, but with the halt of acceptance the protocol clock also halts. That means that timestamps are not reliable anymore. 

> seems to be a patch that is not coherent with the flow of events.
It is the natural response of a validator. If you've been offline and everybody else has been offline too, then you pretend that you did your work to get the rewards. 


> only a marginal number of blocks should be affected

The problem is not how many blocks might need to be unaccepted, the problem is the fact that it might be necessary. 

> Additionally, this solution should be fairly simple to implement.

On the contrary. Unaccepting blocks means reverting state in all components that keep track of blocks and the state of a slot until it is being committed. That means ledger state and potentially depending transactions (very complicated), tangle root, attestations, accounts, rewards/performance factor, etc. This adds an immense amount of complexity. 

I'd definitely prefer something like solution 1a if we can't go the backdating route.

**Comment by vekkiokonio** [comment]
Makes sense. I am still skeptical about backdating but I have no real counter arguments. Considering we're dealing with a corner case (which should happen with very little probability), I guess the simplest solution may simply be the best one at this stage. We can have this as a task to research in the next future.

**Comment by cyberphysic4l** [comment]
Update from meeting on 20/09:

The difficult part of the above proposals is that we were trying to ensure that all accepted blocks end up with a future cone. However, this is not necessary, and there are other cases where accepted blocks don't end up with a future cone. Instead, we should take the commitment to a slot as an implicit reference to all the blocks within, so if we commit to a slot, then everything in the slot automatically has a future cone and we don't need to worry about referencing every single accepted block in the slot. 

Details need to be discussed with a broader iota-core group to ensure the syncing and solidification process required for this solution is acceptable. 

The question of whether validators should backdate blocks remains open, and it is not a necessary step to include more tips anymore. Instead, it is just a question of whether a validator would do this backdating anyway in order to get rewards for the past slots rather than committing to empty slots.
